apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,336 @@
+        });
+      });
+    });
+  }
+
+  private async _runCopyFilesForHeftEvent(
+    heftEvent: HeftEvent,
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration
+  ): Promise<void> {
+    const eventActions: IHeftEventActions = await CoreConfigFiles.getConfigConfigFileEventActionsAsync(
+      logger.terminal,
+      heftConfiguration
+    );
+
+    const copyConfigurations: IExtendedSharedCopyConfiguration[] = [];
+    for (const copyFilesEventAction of eventActions.copyFiles.get(heftEvent) || []) {
+      copyConfigurations.push(...copyFilesEventAction.copyOperations);
+    }
+
+    await this.runCopyAsync({
+      buildFolder: heftConfiguration.buildFolder,
+      copyConfigurations,
+      logger,
+      watchMode: false
+    });
+  }
+
+  protected async runCopyAsync(options: ICopyFilesOptions): Promise<void> {
+    const { logger, buildFolder, copyConfigurations } = options;
+
+    const startTime: number = performance.now();
+    const copyDescriptors: ICopyFileDescriptor[] = await this._getCopyFileDescriptorsAsync(
+      buildFolder,
+      copyConfigurations
+    );
+
+    if (copyDescriptors.length === 0) {
+      // No need to run copy and print to console
+      return;
+    }
+
+    const { copiedFileCount, linkedFileCount } = await this.copyFilesAsync(copyDescriptors);
+    const duration: number = performance.now() - startTime;
+    logger.terminal.writeLine(
+      `Copied ${copiedFileCount} file${copiedFileCount === 1 ? '' : 's'} and ` +
+        `linked ${linkedFileCount} file${linkedFileCount === 1 ? '' : 's'} in ${Math.round(duration)}ms`
+    );
+
+    // Then enter watch mode if requested
+    if (options.watchMode) {
+      await this._runWatchAsync(options);
+    }
+  }
+
+  protected async copyFilesAsync(copyDescriptors: ICopyFileDescriptor[]): Promise<ICopyFilesResult> {
+    if (copyDescriptors.length === 0) {
+      return { copiedFileCount: 0, linkedFileCount: 0 };
+    }
+
+    let copiedFileCount: number = 0;
+    let linkedFileCount: number = 0;
+    await Async.forEachLimitAsync(
+      copyDescriptors,
+      Constants.maxParallelism,
+      async (copyDescriptor: ICopyFileDescriptor) => {
+        if (copyDescriptor.hardlink) {
+          const hardlinkPromises: Promise<void>[] = copyDescriptor.destinationFilePaths.map(
+            (destinationFilePath) => {
+              return FileSystem.createHardLinkAsync({
+                linkTargetPath: copyDescriptor.sourceFilePath,
+                newLinkPath: destinationFilePath,
+                alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+              });
+            }
+          );
+          await Promise.all(hardlinkPromises);
+
+          linkedFileCount++;
+        } else {
+          // If it's a copy, we will call the copy function
+          if (copyDescriptor.destinationFilePaths.length === 1) {
+            await FileSystem.copyFileAsync({
```

Comment:
What ensures that the target folder exists?

Reply:
`fsx.copy` does that internally.